pkg/cloudprovider/provider/aws/provider.go

```diff
@@ -466,19 +466,21 @@ func (p *provider) Create(machine *v1alpha1.Machine, data *cloud.MachineCreateDe
 	}
 	awsInstance := &awsInstance{instance: runOut.Instances[0]}
 
-	// Change to our security group
-	_, err = ec2Client.ModifyInstanceAttribute(&ec2.ModifyInstanceAttributeInput{
-		InstanceId: runOut.Instances[0].InstanceId,
-		Groups:     aws.StringSlice(config.SecurityGroupIDs),
-	})
-	if err != nil {
-		_, err := ec2Client.TerminateInstances(&ec2.TerminateInstancesInput{
-			InstanceIds: []*string{runOut.Instances[0].InstanceId},
+	if len(config.SecurityGroupIDs) > 0 {
+		// Change to our security group
+		_, modifyInstanceErr := ec2Client.ModifyInstanceAttribute(&ec2.ModifyInstanceAttributeInput{
+			InstanceId: runOut.Instances[0].InstanceId,
+			Groups:     aws.StringSlice(config.SecurityGroupIDs),
 		})
-		if err != nil {
-			return nil, awsErrorToTerminalError(err, fmt.Sprintf("failed to attach instance %s to security group id's %v & delete the created instance", aws.StringValue(runOut.Instances[0].InstanceId), config.SecurityGroupIDs))
+		if modifyInstanceErr != nil {
+			_, err := ec2Client.TerminateInstances(&ec2.TerminateInstancesInput{
```

Comment:
it is not clear to my why do we have to delete an instance when adding a security group fails ?

@mrIncompetent said that the nodes cannot communicate if the are no security groups. If we will leave that code as is then perhaps we could generate an `event` of type warning ?

Reply:
We don't have any reconciling otherwise. 
If it fails, the user would get an instance, which does not fit the spec (In this case it might cannot communicate to other instances - Breaking all pod/service communication inside the cluster).

Reply:
may i know why we don't have reconciliation ? it requires too much changes in the code ? it's not practical ?

Reply:
Because most settings of a machine can not be changed after creation, e.G. distro, AMI, VPC. That is why the webhook currently forbids any change to a machine.Spec. However there are a few settings that can only be set _after_ the instance was created as this one which requires us to go this route.

Reply:
right, thanks!